lib/frederick_api/v2/helpers/paginator.rb

```diff
@@ -47,17 +54,54 @@ def current_page
           params.fetch("page.#{page_param}", 1).to_i
         end
 
-        private
-
-          # refetching the next link if no response is found.
-          def next_result_set(current_result_set)
-            if current_result_set
-              response = current_result_set.pages.next
-              response.present? ? response : current_result_set.pages.next
-            else
-              self.result_set.pages.next
+        def is_campaign_source?
+          uri = result_set.links.link_url_for('first')
+          first_params = params_for_uri(uri)
+          filter_string = first_params.fetch('filter.filters')
+          filters_array = JSON.parse(filter_string).flatten
+          filters_array.each do |filter_hash|
+            if filter_hash['operator'] == 'has_no_interaction' &&
+                filter_hash['interaction_type'] == 'delivered_email' &&
+                filter_hash['source_type'] == 'Campaign'
+              return true
             end
           end
+
+          false
+        end
+
+        def eligible_page_count
+          return (total_pages - current_page) unless FrederickAPI.config.emails_per_day_limit_enabled
+
+          url = first_link
+          location_id = url.match(%r{locations/([a-f0-9\-]+)/contacts})[1]
+          cache_key = "emails_sent_today_#{location_id}"
+          emails_sent_today = Rails.cache.read(cache_key) || 0
+          emails_per_day_limit = FrederickAPI.config.emails_per_day_limit
+          batch_size = FrederickAPI.config.frolodex_batch_fetch_size || 1000
+          (emails_per_day_limit - emails_sent_today) / batch_size
+        rescue => e
+          NewRelic::Agent.notice_error(e, first_link: first_link)
+        end
+
+        def pages_to_be_fetched
+          if FrederickAPI.config.jsonapi_campaign_check_enabled && is_campaign_source?
+            [total_pages - current_page, eligible_page_count + 1].min
+          else
+            total_pages - current_page
+          end
+        end
+
+        # log pages fetched for further analysis.
+        def nr_log_page_count(page_count)
+          NewRelic::Agent.record_metric('FrolodexPageFetchCount', page_count)
```

Comment:
metric name can be as per https://docs.newrelic.com/docs/apm/agents/ruby-agent/api-guides/ruby-custom-metrics/#metric_names

Also, is this library has dependency on `new_relic` gem?

Reply:
nope...added rescue for the same reason

Reply:
In this case `.gemspec` should have the required dependency listed.